src/supervisor.adoc

```diff
@@ -1217,30 +1217,60 @@ Two schemes to manage the A and D bits are permitted:
 
 * When a virtual page is accessed and the A bit is clear, or is written
 and the D bit is clear, a page-fault exception is raised.
-* When a virtual page is accessed and the A bit is clear, or is written
-and the D bit is clear, the implementation sets the corresponding bit(s)
-in the PTE. The PTE update must be atomic with respect to other accesses
-to the PTE, and must atomically check that the PTE is valid and grants
-sufficient permissions. Updates of the A bit may be performed as a
-result of speculation, but updates to the D bit must be exact (i.e., not
-speculative), and observed in program order by the local hart.
-Furthermore, the PTE update must appear in the global memory order no
-later than the explicit memory access, or any subsequent explicit memory
-access to that virtual page by the local hart. The ordering on loads and
-stores provided by FENCE instructions and the acquire/release bits on
-atomic instructions also orders the PTE updates associated with those
-loads and stores as observed by remote harts.
-+
-The PTE update is not required to be atomic with respect to the explicit
-memory access that caused the update, and the sequence is interruptible.
-However, the hart must not perform the explicit memory access before the
-PTE update is globally visible.
+* When a virtual page is accessed and the A bit is clear, the PTE is
+  updated to set the A bit.  When the virtual page is written and the D
+  bit is clear, the PTE is updated to set the D bit. When G-stage address
+  translation is in use and is not Bare, the G-stage virtual pages may be
+  accessed or written by implicit accesses to VS-level memory management
+  data structures, such as page tables.                                 +
+                                                                        +
+  When two-stage address translation is in use, an explicit access may
+  cause both VS-stage and G-stage PTEs to be updated. The following rules
+  apply to all PTE updates caused by an explicit or an implicit memory
+  accesses.                                                             +
+                                                                        +
+  The PTE update must be atomic with respect to other accesses to the
+  PTE, and must atomically perform all tablewalk checks for that leaf
+  PTE as part of, and before, conditionally updating the PTE value.
```

Comment:
This appears to say that an implementation must atomically check all intermediate PTEs _in memory_, so on Sv39 an A/D bit update generates a compare-and-compare-and-compare-and-swap bus transaction.  That seems far too strong, is it accurate?

Reply:
The statement is stating "checks for that leaf PTE" and  not all intermediate PTEs. There are no A or   D bits in non-leaf PTEs and the table walker is not required to go revisit any non-leaf PTE that  has been already walked in  the past.

Reply:
Possibly unintended interpretation: "must atomically perform all tablewalk checks (all checks in the table walking process, starting from the root) for that leaf PTE (that would be applicable to any translation which uses that leaf PTE and path)"

Possibly intended interpretation: "must atomically perform all tablewalk checks (all checks performed at the last level only) for that leaf PTE (only the checks which directly concern the PTE itself, not those involved in finding it)"

I see your argument, but I think it requires stretching the meaning of the words a bit.

Reply:
The sentence states "for **that** leaf PTE" and not  for all PTEs leading up to the leaf PTE.

Reply:
I'm not objecting to "that", I'm objecting to "for".  "for" can mean "identified by", and the leaf PTE identifies a set of non-leaf PTEs.

If the clause were changed to "atomically perform all tablewalk checks *acting on* that leaf PTE", it would be perfectly clear.